source/common/http/filter_manager.cc

```diff
@@ -371,6 +371,32 @@ void ActiveStreamDecoderFilter::onDecoderFilterAboveWriteBufferHighWatermark() {
   parent_.filter_manager_callbacks_.onDecoderFilterAboveWriteBufferHighWatermark();
 }
 
+FilterHeadersStatus ActiveStreamDecoderFilter::decodeHeaders(RequestHeaderMap& headers,
+                                                             bool end_stream) {
+  // Each decoder filter instance checks if the request passed to the filter is gRPC
+  // so that we can issue gRPC local responses to gRPC requests. Filter's decodeHeaders()
+  // called here may change the content type, so we must check it before the call.
+
+  is_grpc_request_ = Grpc::Common::isGrpcRequestHeaders(headers);
+  FilterHeadersStatus status = handle_->decodeHeaders(headers, end_stream);
+
+  // Validate filters did not erroneously remove required headers. If they do, send a direct
+  // response.
+  const Http::Status header_status = HeaderUtility::checkRequiredHeaders(headers);
+  if (!header_status.ok()) {
+    ENVOY_STREAM_LOG(debug, "filter={} removed required headers", parent_,
+                     static_cast<const void*>(this));
+    parent_.stream_info_.setResponseFlag(StreamInfo::ResponseFlag::DownstreamProtocolError);
+    const std::string details =
+        absl::StrCat(StreamInfo::ResponseCodeDetails::get().FilterRemovedRequiredHeaders, "{",
+                     header_status.message(), "}");
+    sendLocalReply(Http::Code::ServiceUnavailable, header_status.message(), nullptr, absl::nullopt,
```

Comment:
Technically I think a filter could remove required headers, and have also already sent a local reply. Do we need to protect against this case or is it not worth it? Comment?

Reply:
Oh, right, great point. My impression was if that happened, we would land in this branch https://github.com/envoyproxy/envoy/blob/01c45321514224ec1846e0c1e29dc8760e283871/source/common/http/filter_manager.cc#L766

I will add an integration test for my education / verification.

Reply:
You were right to check this, it broke in this case :) But closing the loop here since the FM check will disappear.